doc/topics/the-event-loop-timers-and-nexttick

```diff
@@ -0,0 +1,174 @@
+is concerned, its happening **synchronously**, but as far as the user is
+concerned, it is happening **asynchronously** because the API of apiCall() was
+written to always be asynchronous.
+
+This philosophy can lead to some potentially problematic situations.  Take this
+snippet for example:
+
+```javascript
+// this has an asynchronous signature, but calls callback synchronously
+function someAsyncApiCall (callback) {
+  callback();
+};
+
+// the callback is called before `someAsyncApiCall` completes.
+someAsyncApiCall(function () {
+
+  // since someAsyncApiCall has completed, bar hasn't been assigned any value
+  console.log('bar', bar); // undefined
+
+});
+
+var bar = 1;
+```
+
+The user defines `someAsyncApiCall()` to have an asynchronous signature,
+actually operates synchronously.  When it is called, the callback provided to
+`someAsyncApiCall ()` is called in the same phase of the event loop
+because `someAsyncApiCall()` doesn't actually do anything asynchronously.  As a
+result, the callback tries to reference `bar` but it may not have that variable
+in scope yet because the script has not been able to run to completion.
+
+By placing it in a nextTick, the script
+still has the ability to run to completion, allowing all the variables,
+functions, etc., to be initialized prior to the callback being called.   It also
+has the advantage of not allowing the event loop to continue.  It may be useful
+that the user be alerted to an error before the event loop is allowed to
+continue.
+
+## process.nextTick() vs setImmediate()
+
+We have two calls that are similar as far as users are concerned, but their
+names are confusing.
+
+* nextTick fires immediately on the same phase
+* setImmediate fires on the following iteration or 'tick' of the event loop
+
+In essence, the names should be swapped.  nextTick fires more immediately than
+setImmediate but this is an artifact of the past which is unlikely to change.
+Making this switch would break a large percentage of the packages on npm.
+Every day more new modules are being added, which mean every day we wait, more
+potential breakages occur.  While they are confusing, the names themselves won't change.
+
+*We recommend developers use setImmediate in all cases because its easier to
+reason about.*
+
+## Two reasons to use nextTick:
+
+1. Allow users to handle errors, cleanup any then unneeded resources, or
+perhaps try the request again before the event loop continues.
+
+2. If you were to run a function constructor that was to, say, inherit from
+`EventEmitter` and it wanted to call an event within the constructor.  You can't
+emit an event from the constructor immediately because the script will not have
+processed to the point where the user assigns a callback to that event.  So,
+within the constructor itself, you can set a callback to emit the event after
+the constructor has finished, which provides the expected results.
```

Comment:
@techjeffharris The event loop doesn't start until after bootstrap. Which includes the initial script passed to node. Here's a partial native stack of running `node -e 'process.abort'`:
```
    frame #2: 0x00000000015e00d1 node_g`node::Abort(args=0x00007fffffffc160) + 17 at node.cc:1643
...
    frame #34: 0x0000000000ccfce1 node_g`v8::Function::Call(this=0x0000000001efe550, recv=(val_ = 0x0000000001efe330), argc=0, argv=0x0000000000000000) + 113 at api.cc:4422
    frame #35: 0x00000000015c7732 node_g`node::Environment::KickNextTick(this=0x0000000001ef2ae0) + 242 at env.cc:81
    frame #36: 0x00000000015d9f1c node_g`node::MakeCallback(env=0x0000000001ef2ae0, recv=(val_ = 0x0000000001efe330), callback=(val_ = 0x0000000001ef0ae8), argc=2, argv=0x00007fffffffd9d0) + 1916 at node.cc:1192
    frame #37: 0x00000000015da166 node_g`node::MakeCallback(env=0x0000000001ef2ae0, recv=(val_ = 0x0000000001efe330), symbol=(val_ = 0x0000000001ef0ad8), argc=2, argv=0x00007fffffffd9d0) + 214 at node.cc:1218
    frame #38: 0x00000000015da1e1 node_g`node::MakeCallback(env=0x0000000001ef2ae0, recv=(val_ = 0x0000000001efe330), method=0x00000000017f7aa6, argc=2, argv=0x00007fffffffd9d0) + 97 at node.cc:1228
    frame #39: 0x00000000015e359d node_g`node::EmitBeforeExit(env=0x0000000001ef2ae0) + 349 at node.cc:3893
    frame #40: 0x00000000015e40af node_g`node::StartNodeInstance(arg=0x00007fffffffdbb0) + 719 at node.cc:4088
    frame #41: 0x00000000015e3d56 node_g`node::Start(argc=1, argv=0x0000000001ea9ad0) + 262 at node.cc:4160
    frame #42: 0x0000000001618f72 node_g`main(argc=3, argv=0x00007fffffffdd18) + 34 at node_main.cc:44
```
If it had been in the event loop then you would have seen `uv_run()`.

Even if you're in the poll phase then execution order of anything queued while _in_ a `setImmediate` callback will be different from anything outside of it. It's always relevant to use `nextTick` in these cases.